find-plugins.js

```diff
@@ -0,0 +1,181 @@
+
+  // Contains package.json but no index.js file?
+  const packageDirent = list.find((dirent) => dirent.name === 'package.json')
+  if (packageDirent && !indexDirent) {
+    throw new Error(`@fastify/autoload cannot import plugin at '${dir}'. To fix this error rename the main entry file to 'index.js' (or .cjs, .mjs, .ts).`)
+  }
+
+  // Otherwise treat each script file as a plugin
+  await processList({ list, opts, indexDirent, prefix, dir, depth, currentHooks, hookedAccumulator })
+
+  return hookedAccumulator
+}
+
+function findCurrentHooks ({ dir, list, hooks, opts, hookedAccumulator, prefix }) {
+  if (!opts.autoHooks) return []
+
+  let currentHooks = []
+  // Hooks were passed in, create new array specific to this plugin item
+  for (const hook of hooks) {
+    currentHooks.push(hook)
+  }
+
+  // Contains autohooks file?
+  const autoHooks = list.find((dirent) => opts.autoHooksPattern.test(dirent.name))
+  if (autoHooks) {
+    const autoHooksFile = join(dir, autoHooks.name)
+    const { type: autoHooksType } = getScriptType(autoHooksFile, opts.packageType)
+
+    // Overwrite current hooks?
+    if (opts.overwriteHooks && currentHooks.length > 0) {
+      currentHooks = []
+    }
+
+    // Add hook to current chain
+    currentHooks.push({ file: autoHooksFile, type: autoHooksType })
+  }
+
+  hookedAccumulator[prefix || '/'].hooks = currentHooks
+
+  return currentHooks
+}
+
+function processIndexDirentIfExists ({ opts, list, dir, hookedAccumulator, prefix }) {
+  // Contains index file?
+  const indexDirent = list.find((dirent) => opts.indexPattern.test(dirent.name))
+  if (!indexDirent) return null
+
+  const file = join(dir, indexDirent.name)
+  const { language, type } = getScriptType(file, opts.packageType)
+  handleTypeScriptSupport(file, language, true)
+  accumulatePlugin({ file, type, opts, hookedAccumulator, prefix })
+
+  const hasDirectory = list.find((dirent) => dirent.isDirectory())
+  if (!hasDirectory) {
+    return hookedAccumulator
+  }
+
+  return indexDirent
+}
+
+async function processList ({ list, opts, indexDirent, prefix, dir, depth, currentHooks, hookedAccumulator }) {
+  const directoryPromises = []
+  for (const dirent of list) {
+    if (opts.ignorePattern && RegExp(opts.ignorePattern).exec(dirent.name)) {
+      continue
+    }
+
+    const atMaxDepth = Number.isFinite(opts.maxDepth) && opts.maxDepth <= depth
+    const file = join(dir, dirent.name)
+    if (dirent.isDirectory() && !atMaxDepth) {
+      processDirectory({ prefix, opts, dirent, dir, file, directoryPromises, hookedAccumulator, depth, currentHooks })
+    } else if (indexDirent) {
+      // An index.js file is present in the directory so we ignore the others modules (but not the subdirectories)
+    } else if (dirent.isFile() && opts.scriptPattern.test(dirent.name)) {
+      processFile({ file, opts, dirent, hookedAccumulator, prefix })
+    }
+  }
+
+  await Promise.all(directoryPromises)
```

Comment:
Can you avoid having a `directoryPromises` array at all? I think this is one of the things that makes this code very hard to read. We should be making `processDirectory` `async` instead and `await` it in place.

Reply:
Yes.